0012-common-ledger-api/0012-common-ledger-api.md

```diff
@@ -0,0 +1,1181 @@
+
+
+### Account Resource
+[Account resource]: #account-resource
+
+An account resource contains the fields in the following table. Some fields are _Server-provided_, meaning they cannot be set directly by clients. Fields that are "Optional" or "Server-provided" in the following table may be omitted by clients submitting objects to the API, but those fields are not optional to implement. Fields of nested objects are indicated with a dot (`.`) character; no fields contain a dot literal.
+
+| Name                      | Type    | Description                            |
+|:--------------------------|:--------|:---------------------------------------|
+| `name`                    | String  | Name of the account. A ledger MAY require this to be unique. MUST match the regular expression `^[a-zA-Z0-9._~-]{1,256}$`. |
+| `fingerprint`             | String  | _(Optional)_ A fingerprint of the account's client certificate. This field MUST be available if and only if the ledger supports client certificate authentication. MUST match the regular expression `^[0-9A-Fa-f]{2}(:[0-9A-Fa-f]{2}){19,127}$`. |
+| `ledger`                  | [URL][] | _(Optional)_ The path to the ledger containing this account. MUST be an HTTP(S) URL where the client can [Get Ledger Metadata][]. |
+| `minimum_allowed_balance` | String  | _(Optional)_ The minimum balance permitted on this account. The special value `"-infinity"` indicates no minimum balance. This is a string so that no precision is lost in JSON encoding/decoding. The default value SHOULD be `"0"`. |
+| `id`                      | [URL][] | _(Server-provided)_ The primary, unique identifier for this account. MUST be an HTTP(S) [URL][] where the client can [get this resource][Get Account]. |
+| `balance`                 | String  | _(Server-provided)_ Balance as decimal [amount][]. Defaults to `"0"`. This can be negative, if the account's `minimum_allowed_balance` allows it. |
+
+A ledger MAY define additional fields for the account resource. For example, `is_admin` could be a boolean that, if true, grants administrator-level authorization when authenticated as the account owner.
+
+
+### Message Resource
+[Message resource]: #message-resource
+
+Messages are sent through the ledger's [Send Message][] method and received in a [WebSocket][] subscription. All fields of the message are required:
+
+| Field    | Value   | Description                                             |
+|:---------|:--------|:--------------------------------------------------------|
+| `ledger` | [URL][] | The base [URL][] of this ledger. MUST be an HTTP(S) URL where the client can [Get Ledger Metadata][]. |
```

Comment:
Even if this saves a couple lines of code in the common use case, that doesn't necessarily mean it is simpler for consumers. I think it's best to design by principle and avoid the temptation to violate the principle to achieve small conveniences. In this case the principle is that a data structure should have a single purpose and only require the minimal amount of information needed to achieve that purpose. There are two purposes here: (1) execute a transfer on a ledger (2) specify a transfer in a way that can be communicated globally. The principle would suggest that the second would have a data structure like `{ledger: ledgerURL, transfer: transfer}`. If we collapse these two data structures into one, we force a burden on code (either client or server) that only cares about the first purpose, and also potentially cause confusion about why the "ledger" field needs to be there. It's hard to list all the disadvantages of "convenience hacks" and weigh them against the convenience because most of the disadvantages are in the higher-level design that it imposes vs the more obvious and low-level convenience, but high-level design is much more important than low-level convenience, which is why it's generally an easy decision to stick with the principles.